File: hydrof/hydro.f90
!
! F77-ized version of a Fortran90 TVD split MHD code
!
!  Original F90 code  copyright (C) 2001,2003, Ue-Li Pen
!  written November 2001 by Ue-Li Pen, pen@cita.utoronto.ca
!
! see http://arxiv.org/abs/astro-ph/0305088
! or http://www.cita.utoronto.ca/~pen/MHD
!
! This code is licencensed under the GPL
!
program hydro
  use ppm, only : outputppm
  use solver, only : nvars, timestep, idens, initialconditions, nguard
  use plot, only : openplot, showplot, closeplot

  implicit none
  integer :: n,nx,ny
  real  :: t,dt
  integer :: iter
  real, allocatable, dimension(:,:,:) :: u
  character(len=30) :: arg

  ! get domain size from command line; if nothing
  ! on command line, prompt user

  if (command_argument_count() < 1) then
      print *,'Enter size of domain, in pixels '
      do 
          read *, n
          if (n < 2 .or. n>500) then
            print *, 'invalid n = ', n, ' try again.'
          else
            exit
      endif
      enddo 
  else
      call get_command_argument(1, arg)
      read( arg,'(I30)'), n
      if (n < 2 .or. n > 500) then
         print *,'invalid n = ', trim(arg), ' using 100'
         n = 100
      endif
  endif
  nx = n+2*nguard   ! boundary condition zones on each side
  ny = n+2*nguard
  allocate(u(nvars,nx,ny))

  call initialconditions(u)
  call outputppm(u,'ics.ppm',idens)
  call openplot(nx, ny)
  t=0
  timesteps: do iter=1,nx*12
      call timestep(u,dt)
      t = t + 2*dt
      if (mod(iter,10) == 1) then 
        print *, iter, 'dt = ', dt, ' t = ', t
        call showplot(u)
      endif 
  end do timesteps
  call outputppm(u,'dens.ppm',idens)

  deallocate(u)
end

File: hydroc/hydro.c
/*  Original F90 code  copyright (C) 2001,2003, Ue-Li Pen
 *  written November 2001 by Ue-Li Pen, pen@cita.utoronto.ca
 *
 * see http://arxiv.org/abs/astro-ph/0305088
 * or http://www.cita.utoronto.ca/~pen/MHD
 *
 * This code is licencensed under the GPL
 */

#include <stdio.h>
#include <stdlib.h>
#include "ppm.h"
#include "plot.h"
#include "solver.h"
#include "allocmultid.h"

int main(int argc, char **argv) {
  int   n, nx, ny;
  float t, dt;
  int   iter;
  float ***u;

  if (argc != 2) {
      printf("Usage: %s n, where n is size of domain.\n", argv[0]);
      exit(-1);
  } else {
      n = atoi(argv[1]);
      if (n < 2 ||  n > 500) {
         n = 100;
         printf("%s: invalid n = %s; using %d.\n", argv[0], argv[1], n);
      }
  }

  nx = n+4; /* two cells on either side for BCs */
  ny = n+4;
  u = alloc3d_float(ny,nx,NVARS);

  initialconditions(u, nx, ny);
  outputppm(u,nx,ny,NVARS,"ics.ppm",IDENS);
  t=0.;
  for (iter=0; iter < 6*nx; iter++) {
      timestep(u,nx,ny,&dt);
      t += 2*dt;
      if ((iter % 10) == 1) {
        printf("%4d dt = %f, t = %f\n", iter, dt, t);
        plot(u, nx, ny);
      }
  }
  outputppm(u,nx,ny,NVARS,"dens.ppm",IDENS);
  closeplot();

  free3d_float(u,ny);
  return 0;
}
